File: algorithms/05_csharp/graphs/cycle_detection.cs
using System;
using System.Collections.Generic;

namespace Algorithms.Graphs
{
    public static class CycleDetection
    {
        public static bool HasCycle(int n, List<(int from, int to)> edges)
        {
            var graph = new List<int>[n];
            for (int i = 0; i < n; i++) graph[i] = new List<int>();
            foreach (var (u, v) in edges) graph[u].Add(v);

            var state = new int[n];
            for (int i = 0; i < n; i++)
            {
                if (state[i] == 0 && Dfs(i, graph, state))
                    return true;
            }
            return false;
        }

        private static bool Dfs(int u, List<int>[] graph, int[] state)
        {
            state[u] = 1;
            foreach (int v in graph[u])
            {
                if (state[v] == 1) return true;
                if (state[v] == 0 && Dfs(v, graph, state)) return true;
            }
            state[u] = 2;
            return false;
        }

        public static void Main()
        {
            var withCycle = new List<(int, int)> { (0, 1), (1, 2), (2, 0) };
            var acyclic = new List<(int, int)> { (0, 1), (1, 2), (0, 2) };
            bool ok = HasCycle(3, withCycle) && !HasCycle(3, acyclic);
            Console.WriteLine(ok ? "[C# CycleDetection] ... verified" : "[C# CycleDetection] ... FAILED");
        }
    }
}

File: algorithms/08_r/graphs/cycle_detection.R
cycle_detection <- function(n, edges) {
  graph <- vector("list", n)
  for (i in 1:n) graph[[i]] <- integer(0)
  for (k in seq_len(nrow(edges))) {
    u <- edges$from[k]
    v <- edges$to[k]
    graph[[u]] <- c(graph[[u]], v)
  }
  state <- integer(n)  # 0 unvisited, 1 visiting, 2 done

  dfs <- function(u) {
    state[u] <<- 1
    for (v in graph[[u]]) {
      if (state[v] == 1) return(TRUE)
      if (state[v] == 0 && dfs(v)) return(TRUE)
    }
    state[u] <<- 2
    FALSE
  }

  for (i in 1:n) {
    if (state[i] == 0 && dfs(i)) return(TRUE)
  }
  FALSE
}

# Example
with_cycle <- data.frame(from = c(1, 2, 3), to = c(2, 3, 1))
acyclic <- data.frame(from = c(1, 2, 1), to = c(2, 3, 3))
stopifnot(cycle_detection(3, with_cycle))
stopifnot(!cycle_detection(3, acyclic))
print("[R cycle_detection] ... verified")
